WordPress/gutenberg
# Retain the same specificity for non iframed selectors

## What?
Alternative to #56649
Fixes #63925

Tries to maintain specificity across styles between the iframed and non-iframed canvas. This PR is slightly different to https://github.com/WordPress/gutenberg/pull/56649 in that it also seeks to maintain specificity for root selectors.

For a quick recap, in a a non-iframed canvas `.editor-styles-wrapper` is prefixed onto the majority of selectors to scope the styles. This results in a bump in specificity. For root selectors (any that begin with `body`, `html` or `:root`) prefixing the classname would create an invalid selector, so the root part of the classname is replaced instead. This results in some of these selectors maintaining specificity, while others change ever so slightly.

This inconsistency creates a situation where the specificity of some styles is bumped, others aren't, and so in a non-iframed editor styles often don't apply in the same way and can look completely different to the front end or iframed editor.

## How?
For most selectors, the prefix is changed from `.editor-styles-wrapper` to `:where(.editor-styles-wrapper)`.

For root selectors, a selector like `body .some-selector` changes to `body :where(.editor-styles-wrapper) .some-selector`. 

Both are maintaining the same specificity.

The tool of choice for prefixing is `postcss` and in particular a `postcss` plugin called `postcss-prefixwrap` is used in trunk. In this PR I've replaced it with the `postcss-prefix-selector` plugin. This one has a callback that allows us to change how scoping works on a per selector basis, which is how the root selector behavior has been changed.

One drawback of the new library is that it's `exclude` option doesn't work the same way as the `ignoredSelectors` option of the old library. Fortunately the callback means it's possible to replicate that behavior.

## Testing Instructions
This needs a lot of general smoke testing. Below are instructions to check whether #63925 was fixed:

[Test using WordPress playground](https://playground.wordpress.net/#%7B%22landingPage%22:%22/wp-admin/%22,%22steps%22:%5B%7B%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22%7D,%7B%22step%22:%22mkdir%22,%22path%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22writeFile%22,%22path%22:%22/wordpress/pr/pr.zip%22,%22data%22:%7B%22resource%22:%22url%22,%22url%22:%22/plugin-proxy.php?org=WordPress&repo=gutenberg&workflow=Build%2520Gutenberg%2520Plugin%2520Zip&artifact=gutenberg-plugin&pr=64534%22,%22caption%22:%22Downloading%20Gutenberg%20PR%2064534%22%7D,%22progress%22:%7B%22weight%22:2,%22caption%22:%22Applying%20Gutenberg%20PR%2064534%22%7D%7D,%7B%22step%22:%22unzip%22,%22zipPath%22:%22/wordpress/pr/pr.zip%22,%22extractToPath%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22installPlugin%22,%22pluginZipFile%22:%7B%22resource%22:%22vfs%22,%22path%22:%22/wordpress/pr/gutenberg.zip%22%7D%7D%5D%7D)

1. Go to Tools > Theme File Editor > theme.json
2. Add the following snippet to the theme.json under `styles.blocks`:
```json
"core/heading": {
    "typography": {
        "fontSize": "var(--wp--preset--font-size--small)"
    }
},
```
3. Create a post and insert a heading block - observe it should have a small font size
4. Save the post and in the editor preferences, enable Custom fields to un-iframe the editor
5. When the editor reloads, observe that the heading block still has a small font size

Here are some diffs that show the styles before and after transformation:
- [Twenty Seventeen Diff](https://gist.github.com/talldan/81ec7b125adc394bf5d86db93ca02447)
- [Twenty Twenty One Diff](https://gist.github.com/talldan/26b4739cceb1cedc9f2091357c3e37c1)
- [Twenty Twenty Four Diff](https://gist.github.com/talldan/806258c68ed0763cb22a3ea5cd155e25)
